unstructured/partition/html/parser.py

```diff
@@ -216,46 +371,48 @@ def _element_from_text_or_tail(
 
         Note this mutates `q` by popping phrasing elements off as they are processed.
         """
-        text_segments = tuple(self._iter_text_segments(text, q))
-        normalized_text = " ".join("".join(ts.text for ts in text_segments).split())
+        element_accum = self._element_accum
 
-        if not normalized_text:
-            return
+        for node in self._iter_text_segments(text, q):
+            if isinstance(node, TextSegment):
+                element_accum.add(node)
+            else:
+                # -- otherwise x is an Element, which terminates any accumulating Element --
+                yield from element_accum.flush(ElementCls)
+                yield node
 
-        # -- if we don't have a more specific element-class, choose one based on the text --
-        if ElementCls is None:
-            ElementCls = derive_element_type_from_text(normalized_text)
-            # -- normalized text that contains only a bullet character is skipped --
-            if ElementCls is None:
-                return
-            # -- derived ListItem means text starts with a bullet character that needs removing --
-            if ElementCls is ListItem:
-                normalized_text = clean_bullets(normalized_text)
-                if not normalized_text:
-                    return
+        yield from element_accum.flush(ElementCls)
 
-        category_depth = self._category_depth(ElementCls)
+    def _iter_text_segments(
+        self, text: str, q: deque[Flow | Phrasing]
+    ) -> Iterator[TextSegment | Element]:
+        """Generate zero-or-more `TextSegment`s or `Element`s from text and leading phrasing.
 
-        yield ElementCls(
-            normalized_text,
-            metadata=ElementMetadata(
-                **_consolidate_annotations(text_segments), category_depth=category_depth
-            ),
-        )
+        Note that while this method is named "._iter_text_segments()", it can also generate
```

Comment:
Pretty sure you told me this earlier too, but are we processing the invalid HTML and saying "quality of the results may be impacted if the HTML is invalid"?

Reply:
No. I tried that approach first, but it just wasn't right. We were changing the "paragraph-ness" of the text in that approach and it was inconsistent with how the browser treated that situation.

So it was a little harder, but now we match what the browser does.

This is a pretty good example.
https://github.com/Unstructured-IO/unstructured/blob/2a12131b4e452b40e1597be888dcfb46c882f9aa/test_unstructured/partition/html/test_parser.py#L688-L700

`<strong>` is phrasing and not meant to have a block item nested in it (according to the HTML Standard). However, when it does, the browser treats the text before the block (`<p>` in this case) as phrasing and adds it to the end of the current paragraph. Then it creates a new paragraph for the block item, and then starts a new paragraph for the phrasing after the block item.

Reply:
Gotcha, thanks. And yeah let's stick with what the browser does.